rfd/0033-sni-alpn-teleport-proxy-routing.md

```diff
@@ -0,0 +1,221 @@
+    alice@node.example.teleport.sh
+```
+
+
+#### tsh db connect db-instance
+`tsh db connect` command introduced in [#7213]([https://github.com/gravitational/teleport/pull/7213) will be extended. After detecting that cluster proxy support SNI ALPN routing
+the `db connect` will start on-demand local proxy and connect to the database through the local proxy using one psql, mysql or mongo shells. When a user exits from the db cli shell a local proxy will be automatically terminated.
+
+```
+                                                             
+         ┌─────────────┐                         
+         │   DB CLI    │                         
+         └─┬────────▲──┘                         
+           │        │                            
+       1) DB     12) DB                          
+      Protocol  Protocol                         
+           │        │                            
+         ┌─▼────────┴──┐                         
+         │ Local Proxy │                         
+         └──┬────────▲─┘                         
+            │        │                           
+       2) TLS SNI 11) TLS                        
+            │        │                           
+            │        │                           
+         ┌──▼────────┴─┐                         
+         │Cluster Proxy│                         
+         └──┬────────▲─┘                         
+            │        │                           
+        3) DB     10) DB                         
+       Protocol  Protocol                        
+            │        │                           
+         ┌──▼────────┴─┐                         
+         │  DB Proxy   │                         
+         │   Service   │                         
+         └──┬────────▲─┘                         
+            │        │                           
+         4) DB    9) DB                          
+        Protocol Protocol                        
+            │        │   5) Auth                 
+         ┌──▼────────┴─┐   Req    ┌─────────────┐
+         │             ├──────────▶             │
+         │ DB Service  │ 6) Auth  │    Auth     │
+         │             ◀───Resp───┤             │
+         └──┬────────▲─┘          └─────────────┘
+            │        │                           
+         7) DB      8) DB                        
+        Protocol   Protocol                      
+            │        │                           
+         ┌──▼────────┴─┐                         
+         │ DB Instance │                         
+         │             │                         
+         └─────────────┘
+```
+
+#### tsh proxy db start db-protocol [-p port]
+
+In order to support connection for standalone DB clients the `tsh proxy db db-protocol` command will provide ability to start run local proxy manually allowing external DB clients to connect to cluster proxy TLS port through the local proxy.
```

Comment:
I think that instead of starting a proxy per protocol, we need to start proxy per database. For example, if there's a Postgres "example" database, a user would do

```
tsh db login example
tsh proxy db example
psql "host=<local-proxy> dbname= user="
```

This will mean the proxy knows which particular database service it's serving and allow it to inject relevant information in the connection which is needed for example to fix [this issue](https://github.com/gravitational/teleport/issues/6584).